deep_qa/data/vocabulary.py

```diff
@@ -0,0 +1,111 @@
+from collections import defaultdict
+import codecs
+import logging
+
+import tqdm
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+class Vocabulary:
+    """
+    A Vocabulary maps strings to integers, allowing for strings to be mapped to an
+    out-of-vocabulary token.
+
+    Vocabularies are fit to a particular dataset, which we use to decide which tokens are
+    in-vocabulary.
+
+    Vocabularies also allow for several different namespaces, so you can have separate indices for
+    'a' as a word, and 'a' as a character, for instance.  Most of the methods on this class allow
+    you to pass in a namespace; by default we use the 'tokens' namespace, and you can omit the
+    namespace argument everywhere and just use the default.
+    """
+    def __init__(self):
+        self._padding_token = "@@PADDING@@"
+        self._oov_token = "@@UNKOWN@@"
+        self._token_to_index = defaultdict(lambda: {self._padding_token: 0, self._oov_token: 1})
+        self._index_to_token = defaultdict(lambda: {0: self._padding_token, 1: self._oov_token})
+        self._finalized = False
+
+    def set_from_file(self, filename: str, oov_token: str="@@UNKNOWN@@", namespace: str="tokens"):
```

Comment:
What I'm currently thinking is to have the vocabulary's constructor look like this:

```python
def __init__(self,                                   
             counter: Dict[str, Dict[str, int]]=None,
             non_padded_namespaces: List[str]=None,  
             min_count: int=1                        
             max_vocab_size: int=None):              
```

(The non-padded namespaces bit is so that we can use the same vocabulary object for handling tag sequences and other labels that need to be mapped to integers, but don't need padding or OOV token ids.)

This way, the vocabulary is given an initial set of counts, and constructs itself that way, and we can have a `from_dataset()` method that computes counts from the dataset, then passes those counts into the constructor.  Does this make sense?  Any better recommendations?